Add School Info Feature

hailed748 opened the pull request:
Purpose:  creates as add school information feature that can be used to populate a database with information regarding a school when members of the school first use the app

Adds: One JavaServlet and one HTML page. The HTML page is temporary as this form can be move to another page in the future

Modifies: This feature is independent of all other features and does not modifying and existing files.


![Screenshot 2020-07-24 at 11 50 02 AM](https://user-images.githubusercontent.com/51325209/88425189-dbd8c000-cda3-11ea-8337-5196ddf25c32.png)

lqueipom commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,171 @@
+   * @return the request parameter, or the default value if the parameter was not specified by the
+   *     client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
```
It doesn't look like you use this function, maybe delete itt?

lqueipom commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,171 @@
+        Entity existingCourse = courseQuery.get(0);
+        existingCourseKey = existingCourse.getKey();
+
+        Entity newTerm = new Entity("Term", existingCourseKey);
```
This would be an existing term right? (consider changing the variable name maybe)

hailed748 replied:
It's actually a new term in all cases because we are assuming you can't have duplicate terms. I didn't make that check, but for safety I will change it such that if a term already exists we won't make a new one.

lqueipom replied:
Perfect! :)

hailed748 commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,171 @@
+      datastore.put(newCourse);
+
+      Entity newProfessor;
+      newProfessor = new Entity("Professor", existingSchoolKey);
```
yup good catch!

ninaprabhu commented on src/main/java/com/google/sps/data/PostRequestObject.java:
```diff
@@ -0,0 +1,38 @@
+  String school;
+  String course;
+  String term;
+  String units;
```
int?

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+          newCourse = new Entity("Course", newSchoolKey);
+        }
+
+        newCourse.setProperty("course-name", courseName);
+        newCourse.setProperty("Units", units);
```
maybe standardize this? so "Course Name" and "Units" or "course-name" and "units" so it's easier for Laura and I to use your database without having to manually refer to each property name

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+
+          newTerm.setProperty("Professor", newProfessorKey);
+          datastore.put(newTerm);
+
```
Nit: Extra line

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+          datastore.prepare(courseQ).asList(FetchOptions.Builder.withDefaults());
+
+      // existing school --> new course
+      if (courseQuery.size() == 0) {
```
use .isEmpty()

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+            datastore.prepare(professorQ).asList(FetchOptions.Builder.withDefaults());
+
+        // existing school --> new course --> new professor
+        if (professorQuery.size() == 0) {
```
.isEmpty()

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+
+        // existing school --> new course --> new professor
+        if (professorQuery.size() == 0) {
+          System.out.println("OLD SCHOOL NEW COURSE NEW PROFESSOR");
```
if this is just for debugging, remove

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+        // existing school --> new course --> new professor
+        if (professorQuery.size() == 0) {
+          System.out.println("OLD SCHOOL NEW COURSE NEW PROFESSOR");
+          Entity newProfessor;
+          newProfessor = new Entity("Professor", existingSchoolKey);
```
Why is this not Entity newProfessor = newEntity("Professor", existingSchoolKey)? I might be missing something but just wondering

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+            datastore.prepare(professorQ).asList(FetchOptions.Builder.withDefaults());
+
+        // existing school --> existing course --> new professor
+        if (professorQuery.size() == 0) {
```
.isEmpty()

ninaprabhu commented on src/main/java/com/google/sps/servlets/addSchoolInfo.java:
```diff
@@ -0,0 +1,179 @@
+
+          newTerm.setProperty("Professor", newProfessorKey);
+          datastore.put(newTerm);
+
```
Nit: Remove line

hailed748 commented:
I made everything lower case for entity names are lower case and properties are lower case as well, hopefully this will standardize things throughout.

ninaprabhu commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+      newTerm.setProperty("term", term);
+      newTerm.setProperty("professor", newProfessorKey);
+      datastore.put(newTerm);
+
```
Nit: extra line? Feel free to ignore this if this is in the style guide - in that case my bad haha

hailed748 replied:
Oh I use an extra line to separate each entity. I think its allowed, kinda a grey area

lqueipom replied:
Agree with Dagmawi on this, makes it a lot easier to read :)

ninaprabhu replied:
Sounds good :)

Tetheta replied:
I also prefer the separate line, clarity is most important here. The main thing we don't like it two separate lines, or a separate line after opening a block via {

ninaprabhu commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+            datastore.prepare(professorQ).asList(FetchOptions.Builder.withDefaults());
+
+        // existing school --> existing course --> new professor
+        if (professorQuery.size() == 0) {
```
.isEmpty()

ninaprabhu commented on src/main/webapp/addSchoolInfo.html:
```diff
@@ -0,0 +1,29 @@
+        <textarea name="school-name">SchoolName</textarea>
+        <br><br>
+        <p>Course Name</p>
+        <textarea name="course-name">CoureName</textarea>
```
Nit: typo

ninaprabhu commented on src/main/webapp/addSchoolInfo.html:
```diff
@@ -0,0 +1,29 @@
+        <textarea name="term">Term</textarea>
+        <br><br>
+        <p>Unit</p>
+        <textarea name="units">Units</textarea>
```
Sidenote: I've been using checkboxes instead of textareas for my units section and I think that makes it easier in the long run (since you won't run into people putting in invalid numbers), maybe consider changing to that?

hailed748 replied:
I changed it to a number input box so they can only input a number from 1 - 24

lqueipom commented on src/test/java/com/google/sps/AddingSchoolInfoTest.java:
```diff
@@ -0,0 +1,54 @@
+//     Assert.assertEquals(testPostObject.getUnits(), mockPost.getParameter("units"));
+//     Assert.assertEquals(testPostObject.getProf(),mockPost.getParameter("professor-name"));
+//   }
+// }
```
Wait, why is this file all commented out?

hailed748 replied:
I need to delete this file I moved it to a different branch

lqueipom commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+    String schoolName = request.getParameter("school-name");
+    String courseName = request.getParameter("course-name");
+    String term = request.getParameter("term");
+    Integer units = Integer.parseInt(request.getParameter("units"));
```
Don't you have to say "int" as type instead of "Integer"? (I remember that I tried to run with Integer and for some reason told me it did not exist)

hailed748 replied:
I have actually always been using Integer this entire internship lol it seems to work

lqueipom replied:
JAJAJA perfect, good that it works for youu :)

Tetheta commented:
Starting review

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+@WebServlet("/add-info")
+public class AddSchoolInfo extends HttpServlet {
+
+  private Key existingSchoolKey;
+  private Key existingCourseKey;
+  private Key existingProfessorKey;
+
+  private Key newSchoolKey;
+  private Key newCourseKey;
+  private Key newProfessorKey;
+  private Key newTermKey;
```
Do these actually need to live for the duration of this Servlet? 
Typically I'd advise keeping parameters close to where they're used and as locally scoped as possible.

hailed748 replied:
I was able to get rid of the new...Key variables but I couldn't figure out how to get rid of the existing key variables

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+    List<Entity> schoolQuery =
+        datastore.prepare(schoolQ).asList(FetchOptions.Builder.withDefaults());
+
+    if (schoolQuery.isEmpty()) {
```
A lot of these if/else statements could be reduced to a function call that encapsulates that switching logic.

This one for instance could be: `getOrCreateSchoolKey(schoolQuery)` and return the school key. Then you've reduced 10 lines into 1 and it's easier to reason about.

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolInfo.java:
```diff
@@ -0,0 +1,159 @@
+      newSchoolKey = null;
+    }
+
+    if (newSchoolKey != null) {
+      Entity newCourse = new Entity("course", newSchoolKey);
+      newCourse.setProperty("course-name", courseName);
+      newCourse.setProperty("units", units);
+      newCourseKey = newCourse.getKey();
+      datastore.put(newCourse);
+
+      Entity newProfessor = new Entity("professor", newSchoolKey);
+      newProfessor.setProperty("professor-name", profName);
+      newProfessorKey = newProfessor.getKey();
+      datastore.put(newProfessor);
+
+      Entity newTerm = new Entity("term", newCourseKey);
+      newTerm.setProperty("term", term);
+      newTerm.setProperty("professor", newProfessorKey);
+      datastore.put(newTerm);
```
You're handling everything at once it seems (course, professor, and term).
While this works, it means there's a lot of code to traverse to get to the details of how the other term creations for example are used and it's easy to miss an edge case.

I'd advise splitting this up into three functions: `getOrCreateCourse`, `getOrCreateProfessor`, and `getOrCreateTerm`.
Then create some booleans if needed like `isNewSchool`, `isNewProfessor`, `isNewTerm` etc. With corresponding functions as needed.

Have the functions pass back the Entity created, then use that for your final data store puts at the end.

Having a conceptual:
- Create course
- Create professor
- Create term
- Store them

With necessary logic aggregated as you go via booleans/etc I think would be clearer.

It's also possible I'm misunderstanding this and you need this extra complexity/branching. If so ping me and we can chat about it further.

Tetheta commented on src/main/webapp/WEB-INF/appengine-web.xml:
```diff
@@ -0,0 +1,10 @@
+  <runtime>java8</runtime>
+  <static-files>
+    <!-- prevent unwanted caching when accessing via the web preview server -->
+    <include path="/**" expiration="0s" />
```
Does this only affect the preview server or will it affect prod as well? Not like it matters here, just curious.

hailed748 replied:
Not really sure how this works. This was a boilerplate given to us in walk-through.

Tetheta replied:
Interesting, guess it doesn't matter.

hailed748 commented:
Adds: refactored servlet now called AddSchoolData, unit test cases under AddSchoolDataTest

Modifies: pom.xml

Removed: Login and Data servlet (these were causing build problems because I was missing some dependencies that Laura and Nina used. I figured i would delete them for now because they will be in the master)

Your suggestion on breaking it up was really good, I tried my best to make it simpler. However I couldn't figure out how to simplify the addSchoolData() function because I need to pass in the parent keys upon entity creation; there is no way to add the parent after creation. Let me know if you have ideas to simplify this.

lqueipom commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,133 @@
+    return result;
+  }
+
+  public void isNewFlagSwitcher(
```
I am having trouble figuring out what the function does by the name, maybe consider changing it?

lqueipom commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+
+  public Entity createSchool(DatastoreService db, String schoolName) {
```
If you initialize datastore at the beginning, do you still need to pass in db as a param?

lqueipom commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,133 @@
+  public Entity createTerm(DatastoreService db, String term, Key professor, Key parent) {
+    Entity newTerm = new Entity("term", parent);
+    newTerm.setProperty("term", term);
+    newTerm.setProperty("professor", professor);
```
It would be good if that variable was called professorKey

ninaprabhu commented on pom.xml:
```diff
@@ -66,7 +107,7 @@
         <configuration>
           <!-- TODO: set project ID. -->
-          <deploy.projectId>nina-laura-dagm-step-2020</deploy.projectId>
+          <deploy.projectId>dagmawi-step-2020</deploy.projectId>
```
Does our project ID not work for your testing? Just curious

hailed748 replied:
I think when I pulled it reverted it

Tetheta replied:
Please change this.

ninaprabhu commented on src/main/webapp/AddSchoolData.html:
```diff
@@ -0,0 +1,29 @@
+    <p>Add A Course Below</p>
+    <form action="/add-info" method="POST">
+        <p>School Name</p>
+        <textarea name="school-name">SchoolName</textarea>
+        <br><br>
+        <p>Course Name</p>
+        <textarea name="course-name">CourseName</textarea>
+        <br><br>
```
Just a small note: I found it more intuitive to leave these text areas blank and give them labels, so we don't have to delete the "SchoolName", "CourseName", existing info every time before adding a course. Not something you need to do, just figured I'd mention that in case you found it easier to work with!

lqueipom commented on src/main/webapp/AddSchoolData.html:
```diff
@@ -0,0 +1,29 @@
+        <br><br>
+        <p>Units</p>
+        <input type="number" name="units" min="1" max="24">
+        <br><br>
```
Consider switching out your bracket tags for padding and margin in CSS (Luke made a comment of this in my PR)

Tetheta commented on .gitignore:
```diff
@@ -1,6 +1,3 @@
-target/
->>>>>>> a18dee48c295115865c85e4418dfb57c05a27f4f
+target/*
+node_modules/*f
```
Why *f?

hailed748 replied:
typo

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,132 @@
+@WebServlet("/add-info")
+public class AddSchoolData extends HttpServlet {
+
+  private Boolean isNewSchool = true;
+  private Boolean isNewCourse = true;
+  private Boolean isNewProfessor = true;
+
+  private Key existingSchoolKey;
+  private Key existingCourseKey;
+  private Key existingProfessorKey;
+
+  private DatastoreService db = DatastoreServiceFactory.getDatastoreService();
```
Same question on the lifecycle of these. Typically you want very few things to live outside of your post/etc function.

Having things live longer can cause issues if you don't reset them all and those bugs are very hard to track down.

hailed748 commented:
@hailed748 ping test?

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,132 @@
+    return newTerm;
+  }
+
+  public void addSchoolData(DatastoreService db, HttpServletRequest request) {
```
Typically we'll put helper functions at the bottom, and the main functions near the top.

That way you can read top to bottom and move as little as possible to see where the code goes.

Also, these should all be private functions.

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,132 @@
+    String profName = request.getParameter("professor-name");
+    Long units = Long.parseLong(request.getParameter("units"));
+
+    isNewEntityDetector(schoolName, courseName, profName);
```
I'd turn this into three small functions that return booleans. That way you don't have to use global variables.

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,132 @@
+    isNewEntityDetector(schoolName, courseName, profName);
+
+    if (isNewSchool) {
+      Entity school = createSchool(schoolName);
```
You don't use these variables for anything, so there's no need to have them.

I'd either advise having your create functions not return anything (since it's not needed), or have them return entities and do a single datastore.put at the end of the entire function. The former is easier here, the latter has the advantage of being very clear about where the storage operation is taking place.

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,132 @@
+
+  private DatastoreService db = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
```
Nice job cleaning this up. It feels a lot more readable now, and hopefully it's also easier to test.
Just a few things and then should be good to go.

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -0,0 +1,449 @@
+
+  @Test
+  public void AddingNewSchool_CreatesNewEntityForAll() {
+    DatastoreService db = DatastoreServiceFactory.getDatastoreService();
+    request = Mockito.mock(HttpServletRequest.class);
+    when(request.getParameter("school-name")).thenReturn("MIT");
+    when(request.getParameter("course-name")).thenReturn("6.006");
+    when(request.getParameter("term")).thenReturn("Spring 2020");
+    when(request.getParameter("units")).thenReturn("12");
+    when(request.getParameter("professor-name")).thenReturn("Jason Ku");
+    newSchoolObject.addSchoolData(db, request);
```
This is used a ton, perfect candidate for creating a helper function.

Something along the lines of `createRequest(schoolName, termName, ...)` 
I think calling out that the db is created and this is added to the datastore is good per test, but repeating these details every time is just noisy.

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -0,0 +1,449 @@
+  @Mock HttpServletRequest requestB;
+
+  @Test
+  public void AddingNewSchool_CreatesNewEntityForAll() {
```
Something we forgot to talk about that would help me a lot when reviewing this.

Typically for each test you'll have three sections: Setup, Act, and Assert.

It may even be useful to call them out explicitly via comment, though typically we just put a blank line between.

What this means for your code here:

Setup is creating the request mock, and any variables needed to check against later (like expected entities/lengths/etc)

Act is actually doing the action, in this case just calling the function on your AddSchoolInfo class.

Assert is anything with assertEquals/etc, where you're verifying the behavior.

You have all the right stuff here, it'd just be easier to read if it was moved into that order.

Tetheta replied:
Also sometimes you might have a pre-assert section where you make sure the DB is empty/etc. Though in this case since you're mocking it out each time I don't think that's necessary.

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -0,0 +1,449 @@
+      String entityType,
+      String entityProperty,
+      String propertyValue,
+      Boolean onlyType) {
```
If this boolean is true, do you nee entityProperty and propertyValue? Seems like you could make those both optional parameters here, or overload your function here for the two different query patterns.

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -0,0 +1,449 @@
+    assertEquals(2, children.size());
+  }
+
+  public List<Entity> findQueryMatch(
```
This doesn't need to be public does it?

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -0,0 +1,449 @@
+import org.mockito.MockitoAnnotations;
+
+/** */
+public final class AddSchoolDataTest {
```
Things I don't see covered yet:

1. What if the data is malformed or missing in the request?
2. What if the datastore.put throws an error?
3. What if the requests are identical?

It's not necessary to handle all these yet, but adding a TODO comment to handle them in the future is typically useful.

hailed748 replied:
I throw an exception now if there is an error on the .put(), and i throw an illegal argument exception if any malformed data is being put. I do not know how to make tests for these though.

hailed748 commented on src/main/webapp/index.html:
```diff
@@ -9,6 +9,7 @@
       src="https://www.gstatic.com/charts/loader.js"
     ></script>
     <script src="script.js"></script>
+    <script src="Graphscript.js"></script>
```
I just changed this. I think you began the review before my other commits went through.

hailed748 commented on src/main/webapp/Graphscript.js:
```diff
@@ -0,0 +1,87 @@
+function getClassID() {
+  var classId = Math.floor(Math.random() * 101);
+  document.getElementById("class-id").innerHTML = classId;
+}
```
I believe this is Nina's code that got pulled when I pulled from master. Likewise with anything course/login related

hailed748 commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -0,0 +1,157 @@
+      course = createCourse(courseName, units, school.getKey());
+      professor = createProfessor(profName, school.getKey());
+      term = createTerm(termName, professor.getKey(), course.getKey());
+      db.put(school);
```
I wasn't quite able to easily create a list so I just went with option 1 and created a function. I will look more into how to easily add multiple things to a list to get the second version working.

Tetheta commented on src/main/java/com/google/sps/servlets/AddSchoolData.java:
```diff
@@ -152,6 +140,28 @@ private Entity createTerm(String term, Key professor, Key parent) {
+      if (professor != null) {
+        db.put(professor);
+      }
+    } catch (DatastoreFailureException | IllegalArgumentException e) {
+      throw e;
+    }
```
Nice job, another advantage of co-locating the puts is you can do something like this.

Tetheta commented on src/main/webapp/graphScript.js:
```diff
@@ -0,0 +1,34 @@
+
+      const difficultyData = new google.visualization.arrayToDataTable(data.difficulty);
+      const diffOptions = {
+          title: "Diffculty of Class",
```
nit: Typo, should be "Difficulty"

Tetheta commented on src/test/java/com/google/sps/servlets/AddSchoolDataTest.java:
```diff
@@ -225,36 +171,20 @@ public void AddingNewTerm_NoNewSchool_NewCourse_NoNewProfessor() {
   @Test
   public void AddingNewTerm_NoNewSchool_NewCourse_NewProfessor() {
```
Just a nit, but one other thing handy to do in tests is make your variables/asserts/etc follow the same order as the test name (and make all test names follow the same order if you have a bunch of essentially boolean logic).

So in this case, declaring expectedTerm, then school, then course, then professor. Then querying & asserting in that same order.
